internal/util/reports.go

```diff
@@ -86,20 +71,26 @@ func (r Reporter) Run(done <-chan struct{}) {
 	}
 }
 
-func (r Reporter) sendStart() {
+func (r *Reporter) sendStart() {
 	signal := prd + "-start"
 	r.send(signal, 0)
 }
 
-func (r Reporter) sendPing(uptime int) {
+func (r *Reporter) sendPing(uptime int) {
 	signal := prd + "-ping"
 	r.send(signal, uptime)
 }
 
-func (r Reporter) send(signal string, uptime int) {
+func (r *Reporter) send(signal string, uptime int) {
 	message := "<14>signal=" + signal + ";uptime=" +
 		strconv.Itoa(uptime) + ";" + r.serializedInfo
-	_, err := r.conn.Write([]byte(message))
+	conn, err := tls.DialWithDialer(&dialer, "tcp", net.JoinHostPort(reportsHost,
+		strconv.FormatUint(uint64(reportsPort), 10)), &tlsConf)
+	if err != nil {
+		r.Logger.Errorf("failed to connect to reporting server: %s", err)
+	}
+	defer conn.Close()
+	_, err = conn.Write([]byte(message))
```

Comment:
From what I can see we're not managing R/W deadlines for this connection, so in theory this goroutine could reach an indefinite hang here under bad networking conditions, or during outage scenarios.

We should configure the write and read deadlines for this connection to avoid that situation, and then add logic to act on the error condition (ideally, by retrying).

Also, we should check the return value of the write to ensure that the number of bytes written matches the length of `message`.

Reply:
Deadline but no retry or length check: if send fails it just gives up. These are low priority and essentially retry by virtue of running every hour anyway: if one fails, wait around for the next ticker to try again.

Reply:
The deadline takes care of the most important thing :+1: